api/client/batch_client.py

```diff
@@ -40,51 +45,68 @@ def get_data(self, url, headers, params=None):
         while retry_count < cfg.MAX_RETRIES:
             start_time = time.time()
             http_request = HTTPRequest(url, method="GET", headers=headers,
-                                       request_timeout=cfg.TIMEOUT, connect_timeout=cfg.TIMEOUT)
-            data = yield self._http_client.fetch(http_request)
-            elapsed_time = time.time() - start_time
-            log_record = dict(base_log_record)
-            log_record['elapsed_time_in_ms'] = 1000 * elapsed_time
-            log_record['retry_count'] = retry_count
-            log_record['status_code'] = data.code
-            if data.code == 200:
+                                       request_timeout=cfg.TIMEOUT,
+                                       connect_timeout=cfg.TIMEOUT)
+            try:
+                data = yield self._http_client.fetch(http_request)
+            except HTTPError as e:
+                elapsed_time = time.time() - start_time
+                log_record = dict(base_log_record)
+                log_record['elapsed_time_in_ms'] = 1000 * elapsed_time
+                log_record['retry_count'] = retry_count
+                log_record['status_code'] = e.code
+                if retry_count < cfg.MAX_RETRIES:
+                    self._logger.warning(e.response.error, extra=log_record)
+                    retry_count += 1
+                    if e.code == 429:
+                        rate_limit_lock.clear()
+                        time.sleep(2 ** retry_count)  # Exponential backoff
+                        rate_limit_lock.set()
+                else:
+                    self._logger.error(e.response.error, extra=log_record)
+                    raise Exception('Giving up on {} after {} tries. \
+                        Error is: {}.'.format(
+                        url, retry_count, e.response.error))
+            else:
```

Comment:
why else? the following block should just be in the try block above

Reply:
Done